---
title: "4.2 Working with Matrices"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)
```

---

## 4.2.1 Logical Indexing

---

### Logical Expressions & Indexing with Matrices

We will continue to work with blood panel data for 4 patients. Remember that the rows are patients and the columns are measurements:

```{r}
pt_A <- c(52, 12.3, 5.2, 6.4, 76)          
pt_B <- c(29, 15.7, 5.8, 9.1, 94) 
pt_C <- c(81, 14.9, 4.3, 5.8, 83)
pt_D <- c(37, 14.3, 54.9, 7.2, 125)
pt_blood <- rbind(pt_A, pt_B, pt_C, pt_D)
colnames(pt_blood) <- c("Age", "HGB", "RBC", "WBC", "BG")
pt_blood
```

When executing a logical expression on a matrix, it returns a *matrix of logical values*. Each element is evaluated individually.

A logical expression with a matrix is equivalent to asking: In this matrix, which values are ______?

For example: In the `pt_blood` matrix, which values are less than 10?

```{r}
pt_blood < 10
```

We can see that the first three values in the RBC column and all values in WBC column are less than 10 as these elements are `TRUE`.

The syntax for indexing a matrix with a matrix of logical indices is:

    matrix_name[logical_matrix]
    
You will notice that there is no comma used in this type of indexing. Much like indexing with matrix element indices, logical matrix indexing returns a vector of values for which the logical matrix has value of `TRUE`:

```{r}
pt_blood[pt_blood < 10]
```

In this case it returns a vector containing the first three values in the RBC column and all the values in the WBC column because these are all less than 10. Why is the result a vector and not a matrix? In many cases, not all rows or columns contain the same number of `TRUE` values. In our logical matrix above, for example, the `RBC` column had three `TRUE` values while the WBC value had four, and you can't have a matrix with unequal row or column lengths.

---

### Logical Expressions & Indexing with Matrix Rows

When executing a logical expression on a *matrix row*, it returns a *vector of logical values*. Each element in the matrix row is evaluated individually.

A logical expression with a matrix row is equivalent to asking: In this matrix row, which values are  ______?

In the case of our `pt_blood` matrix, this is equivalent to: For patient X, which measurements are ______?

For example: For patient A, which measurements are greater than 10?

```{r}
pt_blood["pt_A", ] > 10
```

Notice that we have one logical value for each column, meaning we have a vector of *logical column indices*.

Thus, if we use the results of this logical expression to index our matrix, we need to put it in the place of *column indices*. The syntax for indexing a matrix with a vector of logical column indices is:

    matrix_name[ ,logical_column_indices]
    
Similarly, if you use a logical expression on a row to index a matrix, the syntax is also:

    matrix_name[ ,logical_expression_on_row]
    
To clarify, the difference between these two forms of syntax is that in the first example, we are passing in a pre-determined vector of logicals that correspond to indices for the matrix columns, whereas in the second example, we are performing a logical evaluation on values in the rows that correspond to a particular column or columns. The end result is the same.
    
Let's illustrate this difference with an example. For patient A, Age, HGB and BG values were all greater than 10. We can save the logical expression we wrote above to a variable and then use that variable as our column indices:

```{r}
logical_col_index <- pt_blood["pt_A", ] > 10

# matrix_name[ ,logical_column_indices]
pt_blood[ ,logical_col_index]
```

Or we can skip the step of assigning an intermediate variable by instead using the logical expression directly when indexing the matrix:

```{r}
# matrix_name[ ,logical_expression_on_row]
pt_blood[ ,pt_blood["pt_A", ] > 10]
```

---

### Logical Expressions & Indexing with Matrix Columns

Similarly, when executing a logical expression with a *matrix column*, it returns a *vector of logical values*. Each element in the matrix column is evaluated individually.

A logical expression with a matrix column is equivalent to asking: In this matrix column, which values are ______?

In the case of our `pt_blood` matrix, this is equivalent to: For measurement X, which patients are ______?

For example: For blood glucose (BG), which patients have a level greater than 90?

```{r}
pt_blood[ ,"BG"] > 90
```

Notice that we have one logical value for each row, meaning we have a vector of *logical row indices*. 

Thus, if we use the results of this logical expression to index our matrix, we need to put it in the place of *row indices*. The syntax for indexing a matrix with a vector of logical row indices is:

        matrix_name[logical_row_indices, ]
    
Similarly, if you use a logical expression on a column to index a matrix, the syntax is also:

        matrix_name[logical_expression_on_column, ]
        
To clarify, the difference between these two forms of syntax is that in the first example, we are passing in a pre-determined vector of logicals that correspond to indices for the matrix rows, whereas in the second example, we are performing a logical evaluation on values in the columns that correspond to a particular row or rows. The end result is the same.
    
Let's illustrate this difference with an example. For blood glucose level (BG), patients B and D had values greater than 90. We can save the logical expression we wrote above to a variable and then use that variable as our row indices:

```{r}
logical_row_index <- pt_blood[ ,"BG"] > 90

# matrix_name[logical_row_indices,]
pt_blood[logical_row_index, ]
```

Or we can skip the step of assigning an intermediate variable by instead using the logical expression directly when indexing the matrix:

```{r}
# matrix_name[logical_expression_on_column,]
pt_blood[pt_blood[ ,"BG"] > 90, ]
```
    
Logical row indices and logical column indices can also be combined:

        matrix_name[logical_expression_on_column, logical_expression_on_row]
        
For example, we can determine which patients have a blood glucose measurement greater than 90, and which measurements from patient A have a value greater than 10:

```{r}
pt_blood[pt_blood[ ,"BG"] > 90, pt_blood["pt_A", ] > 10]
```

---

### Logical Expressions & Indexing Summary

**Entire Matrix**
Logical expression on a matrix
Return value: matrix of logical values

Indexing with a matrix of logical values: `matrix_name[logical_matrix]`
Return value: vector of values

**Matrix Row**
Logical expression on a matrix row
Return value: vector of logical column indices

Indexing with a vector of logical column indices: `matrix_name[ ,logical_column_indices]`
Return value: matrix containing all rows, indexed columns

**Matrix Column**
Logical expression on a matrix column
Return value: vector of logical row indices

Indexing with a vector of logical row indices: `matrix_name[logical_row_indices, ]`
Return value: matrix containing indexed rows, all columns

---

## 4.2.2 Matrix Functions

---

### Functions Applied to Each Element and Information Functions

For the rest of this section we will work with expression levels for seven randomly chosen genes across five human tissues (whole blood, liver, lung, pancreas, and stomach). Expression levels are measured in transcripts per million (TPM), where the higher the TPM value, the higher the gene expression in that tissue.

```{r}
blood <- c(7.28, 14.32, 1.67, 0.06, 0.00, 0.01, 8.91)
liver <- c(0.32, 10.87, 0.29, 0.01, 14.98, 0.00, 15.33)
lung <- c(1.36, 31.60, 4.03, 0.04, 0.02, 0.04, 36.50)
pancreas <- c(0.04, 12.02, 0.15, 37.65, 0.13, 0.01, 24.20)
stomach <- c(0.60, 12.17, 0.84, 0.85, 0.00, 0.04, 36.59)
# Create matrix
tissue_TPM <- rbind(blood, liver, lung, pancreas, stomach)
colnames(tissue_TPM) <- c("CD19", "CTCF", "FCER1A", "KIRREL2", "LPA", "RBFOX1", "TUBG1")
tissue_TPM
```

As with vectors, many functions are applied to each individual matrix element. This includes functions like `log()`, `round()`, `sqrt()`, `toupper()`, `tolower()`, `grepl()`, `gsub()`, and `nchar()`.

For example, when we apply `round()` and `log()` to the `tissue_TPM` matrix as below, the natural log of each element will be taken and rounded to two digits, and a matrix with each element transformed is returned:

```{r}
round(log(tissue_TPM), digits = 2)
```

Similarly, many of the functions we learned to provide information about vectors behave the same way with matrices.

The `mean()` function returns the mean of all matrix elements:

```{r}
mean(tissue_TPM)
```

The `sd()` function returns the standard deviation of the matrix elements from the mean:

```{r}
sd(tissue_TPM)
```

The `min()` function returns the lowest value in the matrix:

```{r}
min(tissue_TPM)
```

The `max()` function returns the highest value in the matrix:

```{r}
max(tissue_TPM)
```

---

### Matrix Specific Functions & Matrix Specific Function Behaviour

Some functions are applicable specifically to matrices. The `colMeans()` and `rowMeans()` functions, for example, return a vector of the mean of each column and each row respectively:

```{r}
colMeans(tissue_TPM)
```

```{r}
rowMeans(tissue_TPM)
```

Other functions we have already learned behave differently when applied to matrices. For example, the `summary()` function will not return summary statistics for the entire matrix, but rather for each column in the matrix:

```{r}
summary(tissue_TPM)
```

Another useful matrix-specific function is the transpose function `t()`. Matrix transposition might be a familiar concept to those who have studied linear algebra. Transposing a matrix flips it on the diagonal, or in other words, swaps the rows and columns (row one will become column one, row two will become column two, etc.):

```{r}
tissue_TPM_transposed <- t(tissue_TPM)
tissue_TPM_transposed
```

The transposed version of `tissue_TPM` has genes as rows and tissues as columns. 

The `cor()` function can be used in a different way with matrices than with vectors. Instead of providing two vectors to be correlated, the entire matrix can be used as the only argument, and the correlation between each *pair of columns* will be calculated:

```{r}
tissue_correlations <- round(cor(tissue_TPM_transposed), digits = 2)
tissue_correlations
```

The correlation between the expression of the seven genes in blood and liver is 0.41, as seen at both `[blood, liver]` and `[liver, blood]`. 

The correlation between the expression of the seven genes in pancreas and stomach is 0.42, as seen at both `[pancreas, stomach]` and `[stomach, pancreas]`. 

This is a symmetric 2D matrix, as discussed in Supplement 4.1.1. The rows and columns are the same set of samples, and each value is a measurement of the relationship between the samples.

All of the tissues are perfectly correlated with themselves, so along the diagonal of the matrix all values are 1.

You can check if a matrix is symmetric by seeing if the matrix is equal to the transpose of itself:

```{r}
tissue_correlations == t(tissue_correlations)
```

---

## 4.2.3 R Packages & Heat Maps

---

### R Packages

An R package is a collection of functions and/or datasets that can be installed to add greater functionality to R. To the right you will see a tab called "Packages" that lists all of the packages already installed to the JupyterHub R server.

Some examples of packages you will come across in this course include:
`edgeR`    - Functions for the analysis of RNA sequencing data
`ggplot2`  - Functions for making a wide variety of plots
`pheatmap` - Function for creating well-formatted (pretty) heat maps

In this section we will be learning about heat maps using the `pheatmap` package.

Before you can use a package, it needs to be installed and loaded. **All of the packages you will need to use in this course have already been installed on the server**, but if you need to install packages in the future on a different server or computer, you will need the following functions.

Many R packages, including `pheatmap`, can be installed using the `install.packages()` function, which installs packages stored in the standard R repository CRAN (https://cran.r-project.org/):

    install.packages("pheatmap")

There are numerous biology-specific packages, such as `edgeR`, that are stored separately from the standard R packages at a repository called BioConductor (https://www.bioconductor.org/). To use these packages, you must first install the `BiocManager` package from the standard R repository using the `install.packages()` function:

    install.packages("BiocManager")

And then install the Bioconductor package using the function `BiocManager::install()`:

    BiocManager::install("edgeR")

**Do not worry about installation for this course - this information is provided for your interest and potential future needs.**

Once a package is installed, it must be *loaded* from the *package library* before any of the functions or data sets that come with the package can be used. The `pheatmap` package contains a function called `pheatmap()`. If we try to use the function now, it will not work:

```{r}
pheatmap(tissue_TPM)
```

To load in a package we use the function `library()`. The name of the package, with or without quotes, is the only required argument:

```{r}
library(pheatmap)
# library("pheatmap") also works
```

---

### Heat Maps

A heat map displays 2D numerical data, like the data in a numeric matrix. To better understand heat maps we will start by plotting one to see what they look like using the `pheatmap()` function (do not worry yet about how the `pheatmap()` function works):

```{r}
pheatmap(tissue_TPM, cluster_rows = FALSE, cluster_cols = FALSE)
```

You can see that there is a row for each row in `tissue_TPM` and a column for each column in `tissue_TPM`. Each "cell" (square) has a colour. The colours indicate the magnitude of the value. In this case, the colour represents the TPM of gene X in tissue Y. The legend on the right of the heat map indicates which colour represents which TPM value. 

The cell for the liver row and the RBFOX1 column represents the value 0.00 in the matrix. The cell for the stomach row and the TUBG1 column represents the value 36.59 in the matrix.

In the call to the `pheatmap()` function above, the arguments `cluster_rows` and `cluster_cols` were set to `FALSE`. By default, the rows and columns of heat maps are generally clustered to display trends in the data.

Let's create a heat map with clustering:

```{r}
pheatmap(tissue_TPM)
```

You can see that on the left and top sides of the heat map there are *dendrograms*. A dendrogram is a tree diagram that displays the results of *hierarchical clustering*. Hierarchical clustering is method of clustering data based on the similarity of pieces of data to one another. 

The dendrogram to the left of the plot shows the clustering of the five different tissues based on the similarity of their expression profiles for the seven genes. 

Examining the dendrogram starting closest to the CTCF column, we can see that it groups the blood and the liver tissues together, as well as the lung and the stomach tissues together. Then as we move further left, these four tissues are joined together before finally being joined with the pancreas. 

This indicates:
- blood and liver tissues are more similar to one another than to the other tissues
- lung and stomach tissues are more similar to one another than to the other tissues
- the pancreas is the most dissimilar from the rest of the tissues

Though it may not be perceptible depending on the size of your computer screen, the length of the lines that extend out from the blood and the liver rows are shorter than those than extend out from the lung and stomach rows. This indicates that the blood and liver tissues are more similar to one another than the lung and stomach are to one another.

Looking back at the heat map, we can make some additional observations with respect to these five tissues:
- CTCF and TUBG1 are ubiquitously expressed - they are expressed in all 5 tissue types
- KIRREL2, LPA and CD19 are tissue specific - they are expressed in one tissue type
- FCER1A and RBFOX1 are weakly expressed in these tissues - they are expressed at extremely low levels, if at all, in all five tissue types.

---