---
title: "4.2 Working with Matrices - Tutorial"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, error = TRUE, warning = TRUE)

library(ottr)
library(testthat)
```

---

At the end of this tutorial you should be able to:

* Apply logical expressions to matrix rows and columns
* Use the results of a logical expression to index a matrix
* Apply functions to matrices
* Install and load an R package
* Create a heat map to display the data in a matrix
* Interpret a heat map

---

## 4.2.1: Logical Indexing

---

In this tutorial we will again look at SNP frequencies from the 1000 Genomes Project sourced from dbSNP (https://www.ncbi.nlm.nih.gov/snp/).

We will look at 10 randomly chosen SNPs across 5 populations:
- African (AFR)
- East Asian (EAS)
- European (EUR)
- South Asian (SAS)
- American (AMR)

The matrix of alternative allele frequencies is created below for you to work with throughout the tutorial.

```{r}
rs671 <- c(0.00, 0.17, 0.00, 0.00, 0.00)
rs1384 <- c(0.10, 0.71, 0.49, 0.65, 0.54)
rs3077 <- c(0.58, 0.68, 0.19, 0.37, 0.28)
rs4680 <- c(0.28, 0.28, 0.50, 0.44, 0.32)
rs5093 <- c(0.02, 0.17, 0.03, 0.05, 0.06)
rs6265 <- c(0.01, 0.49, 0.20, 0.20, 0.15)
rs7412 <- c(0.10, 0.10, 0.06, 0.04, 0.05)
rs8362 <- c(0.47, 0.61, 0.30, 0.61, 0.30)
rs9361 <- c(0.06, 0.00, 0.00, 0.00, 0.00)
rs25531 <- c(0.09, 0.22, 0.13, 0.14, 0.05)

SNPs <- cbind(rs671, rs1384, rs3077, rs4680, rs5093, rs6265, rs7412, rs8362, rs9361, rs25531)
rownames(SNPs) <- c("AFR", "EAS", "EUR", "SAS", "AMR")

SNPs
```

Write a logical expression to answer the question: Which populations have an alternative allele frequency greater than 0 for rs9361? Assign the output logical vector to a variable named "rs9361_present".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.1.a.R")
```
**Check your answer by running the above block** 

Use the logical expression written above to output all of the allele frequencies for the population(s) with an alternative allele frequency greater than 0 for rs9361. Assign the output to a variable named "rs9361_present_allele_freqs".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.1.b.R")
```
**Check your answer by running the above block** 

Write a logical expression to answer the question: Which SNPs have an alternative allele frequency greater than 0.05 in the South Asian population? Assign the output logical vector to a variable named "SAS_SNPs_present".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.1.c.R")
```
**Check your answer by running the above block** 

Use the logical expression written above to output, for all populations, the allele frequencies of the SNPs which have an alternative allele frequency greater than 0.05 in the South Asian population.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.1.d.R")
```
**Check your answer by running the above block** 

Output the allele frequencies for: all populations that contain an R, and all SNPs that contain a 2 in the ID. Assign the output matrix to a variable named "pop_R_rsID_2".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.1.e.R")
```
**Check your answer by running the above block** 

---

## 4.2.2: Matrix Functions

---

Use a function to output the mean alternative allele frequency for each SNP. Assign the output to variable named "mean_SNP_AA".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.a.R")
```
**Check your answer by running the above block** 

Use the `summary()` function to get a summary of the alternative allele frequencies for each population. Assign the output to a variable named "SNP_summaries". (Hint: You will need to nest another function inside of the `summary()` function.)
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.b.R")
```
**Check your answer by running the above block** 

Load the pheatmap package (it should already be installed).
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.c.R")
```
**Check your answer by running the above block** 

Create a heat map of alternative allele frequencies. Assign the call to the `pheatmap()` function to a variable named "SNP_heatmap". It will still create a plot.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.d.R")
```
**Check your answer by running the above block** 

Are there any SNPs that seem to have similar frequencies across all 5 populations?

Are there any SNPs that are uniquely absent in one population?

Are there are SNPs that are only found in one population?

Create a matrix of the correlations between each *population* based on the alternative allele frequencies. Call it "SNP_population_correlations".
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.e.R")
```
**Check your answer by running the above block** 

Create a heat map of `SNP_population_correlations`. Assign the call to the `pheatmap()` function to a variable named "population_heatmap". It will still create a plot.
```{r}
# Write and execute your code here

```

```{r}
. = ottr::check("tests/Tutorial.W4.2.2.f.R")
```
**Check your answer by running the above block** 

Which two populations are the most similar to one another based on these SNP allele frequencies?

Which population is the most unique from the rest based on these SNP allele frequencies?

---
